tools/lint-sh.js

```diff
@@ -0,0 +1,182 @@
+#!/usr/bin/env node
+'use strict';
+
+const { execSync, spawn } = require('child_process');
+const { promises: fs, readdirSync, statSync } = require('fs');
+const { extname, join, relative, resolve } = require('path');
+
+const SHELLCHECK_OPTIONS = ['--shell=sh', '--severity=info', '--enable=all'];
+if (process.argv[3] === '--fix') SHELLCHECK_OPTIONS.push('--format=diff');
+else if (process.env.GITHUB_ACTIONS) SHELLCHECK_OPTIONS.push('--format=json');
+
+const SPAWN_OPTIONS = {
+  cwd: null,
+  shell: false,
+  stdio: ['pipe', 'pipe', 'inherit'],
+};
+
+function* findScriptFilesRecursively(dirPath) {
+  const entries = readdirSync(dirPath, { withFileTypes: true });
+
+  for (const entry of entries) {
+    const path = join(dirPath, entry.name);
+
+    if (
+      entry.isDirectory() &&
+      entry.name !== 'build' &&
+      entry.name !== 'changelogs' &&
+      entry.name !== 'deps' &&
+      entry.name !== 'fixtures' &&
+      entry.name !== 'gyp' &&
+      entry.name !== 'inspector_protocol' &&
+      entry.name !== 'node_modules' &&
+      entry.name !== 'out' &&
+      entry.name !== 'tmp'
+    ) {
+      yield* findScriptFilesRecursively(path);
+    } else if (entry.isFile() && extname(entry.name) === '.sh') {
+      yield path;
+    }
+  }
+}
+
+const expectedHashBang = Buffer.from('#!/bin/sh\n');
+async function hasInvalidHashBang(fd) {
+  const { length } = expectedHashBang;
+
+  const actual = Buffer.allocUnsafe(length);
+  await fd.read(actual, 0, length, 0);
+
+  return Buffer.compare(actual, expectedHashBang);
+}
+
+async function checkFiles(...files) {
+  const flags = process.argv[3] === '--fix' ? 'r+' : 'r';
+  await Promise.all(
+    files.map(async (file) => {
+      const fd = await fs.open(file, flags);
+      if (await hasInvalidHashBang(fd)) {
+        if (process.argv[3] === '--fix') {
+          const file = await fd.readFile();
+
+          const fileContent =
+            file[0] === '#'.charCodeAt() ?
+              file.subarray(file.indexOf('\n') + 1) :
+              file;
+
+          const toWrite = Buffer.concat([expectedHashBang, fileContent]);
+          await fd.truncate(toWrite.length);
+          await fd.write(toWrite, 0, toWrite.length, 0);
+        } else {
+          if (!process.exitCode) process.exitCode = 1;
+          console.error(
+            (process.env.GITHUB_ACTIONS ?
+              `::error file=${file},line=1,col=1::` :
+              'Fixable with --fix: ') +
+              `Invalid hashbang for ${file} (expected /bin/sh).`
+          );
+        }
+      }
+      await fd.close();
+    })
+  );
+
+  const stdout = await new Promise((resolve, reject) => {
+    let SHELLCHECK_EXE;
+    const NPX_OPTIONS = [];
+    try {
+      SHELLCHECK_EXE =
+        process.env.SHELLCHECK ||
+        execSync('command -v shellcheck').toString().trim();
+    } catch {
+      SHELLCHECK_EXE = execSync('command -v npx').toString().trim();
+      NPX_OPTIONS.push('shellcheck');
```

Comment:
I'm having second thoughts about this. I'm wondering if, for security purposes, an `npx` fallback should fallback to a specific published version of `shellcheck` (like we do with `dmn` in one of the scripts) to avoid automatically grabbing a version that has been published maliciously. (I don't have any reason to believe the `shellcheck` maintainer would do such a thing, but we've seen stuff happen with maintainer handoffs and things like that in the past....)

Alternatively, we can just remove the `npx` fallback. Yeah, I know, I'm the one that advocated for it (I think) in the first place. Sorry to be indecisive and cause extra work.

Reply:
What do you think about making the `npx` call opt-in? Something like `./tools/lint-sh.js . --from-npx`?

Reply:
That works for me. Thanks.

Reply:
Done, PTAL.